src/librustc_trait_selection/traits/structural_match.rs

```diff
@@ -137,6 +138,10 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
                 self.found = Some(NonStructuralMatchTy::Param);
                 return true; // Stop visiting.
             }
+            ty::Dynamic(..) => {
+                self.found = Some(NonStructuralMatchTy::Dynamic);
+                return true; // Stop visiting.
+            }
```

Comment:
So this looks like we have a fall-through clause here that by default *accepts* all types? Isn't that *incredibly dangerous*, also when new variants are added?
Any match that checks soundness things should be exhaustive, or at the least have a fallback that is safe. Here the fallback evidently was not safe.

Reply:
https://github.com/rust-lang/rust/pull/72153